File: ledger/static/scripts/assets.js
/* Ledger - A web application to track cryptocurrency investments
** Copyright (C) 2021 teleprint.me
**
** This program is free software: you can redistribute it and/or modify
** it under the terms of the GNU Affero General Public License as published by
** the Free Software Foundation, either version 3 of the License, or
** (at your option) any later version.
**
** This program is distributed in the hope that it will be useful,
** but WITHOUT ANY WARRANTY; without even the implied warranty of
** MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
** GNU Affero General Public License for more details.
**
** You should have received a copy of the GNU Affero General Public License
** along with this program.  If not, see <https://www.gnu.org/licenses/>.
*/
const error = (message) => {
    console.log('[Error]', message);
};


const setAssetPromise = (method, url) => {
    let promise = request(method, url);
    promise.then((response) => {
        let assets = JSON.parse(response);
        let datalist = document.querySelector('#pairs');
        removeChildrenFromElement(datalist);
        for (let asset of assets) {
            if (!asset['display']) { continue; }
            let node = document.createElement('option');
            node.setAttribute('value', asset['display']);
            node.setAttribute('data-id', asset['id']);
            node.setAttribute('data-min', asset['min-size']);
            node.setAttribute('data-name', asset['name']);
            datalist.appendChild(node);
        }
    }).catch(error);
    return promise;
};


const setDefaultDatalist = () => {
    let datalist = document.querySelector('#pairs');
    let node = document.createElement('option');
    removeChildrenFromElement(datalist);
    node.setAttribute('value', 'Not Available');
    datalist.appendChild(node);
};


const togglePlatformOnChange = (event) => {
    let platform = event.target.value;
    switch (platform) {
        case 'coinbase-pro':
        case 'kraken':
            setAssetPromise('GET', `/client/${platform}/assets`);
            break;
        default:
            setDefaultDatalist();
            break;
    };
};


const toggleStrategyOnChange = (event) => {
    let apy = event.target.form.apy;
    switch (event.target.value) {
        case 'value-average':
        case 'dynamic-value-average':
            apy.parentElement.classList.remove('d-none');
            apy.setAttribute('required', 'true');
            break;
        case 'cost-average':
        case 'dynamic-cost-average':
        default:
            apy.parentElement.classList.add('d-none');
            apy.removeAttribute('required');
            break;
    }
};


const getMessageOnError = (object) => {
    let message;

    if (!object.platform) {
        message = '[Error] platform field is empty';
    } else if (!object.asset) { 
        message = '[Error] asset field is empty';
    } else if (!object.id) { 
        message = `[Error] failed to set id=${object.id}`;
    } else if (!object.min) {
        message = `[Error] failed to set min=${min}`;
    }

    return message;
};


const getPrincipleData = (datalist) => {
    for (let child of datalist.children) {
        console.log('[Info] seeking `id` and `min`');
        let value = child.getAttribute('value');
        if (value === form.asset.value) {
            let id = child.getAttribute('data-id');
            let min = child.getAttribute('data-min');
            console.log(`[Info] set id=${id} and min=${min} using value=${value}`);
            return {'id': id, 'min': min};
        }
    }
    return {};
};


const togglePrincipleOnFocus = (event) => {
    // grab elements
    let form = event.target.form;
    let datalist = document.querySelector('#pairs');

    // grab required attributes
    let data = getPrincipleData(datalist);

    let fields = {
        'platform': form.platform.value,
        'asset': form.asset.value,
        'id': data.id,
        'min': data.min
    };

    // set message on error
    let message = getMessageOnError(fields);

    // avoid making the request if an error occurred
    if (message) {
        console.log(message);
        event.preventDefault();
        event.stopPropagation();
        return;
    }

    // grab data
    let promise = request('GET', `/client/${fields.platform}/price/${fields.id}`);

    // commit promise
    promise.then((response) => {
        // update principle min value
        let amount = undefined;
        let price = JSON.parse(response);
        switch (fields.platform) {
            case 'coinbase-pro':
            case 'kraken':
                if (fields.min > 0 && fields.min < 1) {
                    amount = Math.ceil(price.ask * fields.min);
                } else {
                    amount = fields.min;
                }
                if (amount <= 5) {
                    amount = 5;
                }
                form.principle.setAttribute('min', amount);
                console.log(`[Info] set amount=${amount} using price['ask']=${price['ask']}`)
                break;
            default:
                form.principle.setAttribute('min', 5);
                break;
        }
    }).catch(error);
};

File: ledger/forms/assets.py
# Ledger - A web application to track cryptocurrency investments
# Copyright (C) 2021 teleprint.me
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU Affero General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Affero General Public License for more details.
#
# You should have received a copy of the GNU Affero General Public License
# along with this program.  If not, see <https://www.gnu.org/licenses/>.
from flask_wtf import FlaskForm

from wtforms.widgets.html5 import NumberInput

from wtforms.fields import SelectField
from wtforms.fields import StringField
from wtforms.fields import SubmitField
from wtforms.fields.html5 import IntegerField

from wtforms.validators import DataRequired
from wtforms.validators import ValidationError

from flask import g


class AssetExists(object):
    def __init__(self, message=None):
        if not message:
            message = 'Asset already exists'
        self.message = message

    def __call__(self, form, field):
        for product in g.db.assets.find():
            has_account = form.platform.data == product['platform']
            has_asset = field.data == product['asset']
            if has_account and has_asset:
                raise ValidationError(self.message)


class AssetsCreateForm(FlaskForm):
    platform = SelectField('Account', [
        DataRequired('Account is required')
    ], choices=[
        ('coinbase', 'Coinbase'),
        ('coinbase-pro', 'Coinbase Pro'),
        ('kraken', 'Kraken')
    ])

    asset = StringField('Asset', [
        DataRequired('Asset is required'),
        AssetExists()
    ])

    strategy = SelectField('Strategy', [
        DataRequired('Strategy is required'),
    ], choices=[
        ('cost-average', 'Cost Average'),
        ('dynamic-cost-average', 'Dynamic Cost Average'),
        ('value-average', 'Value Average'),
        ('dynamic-value-average', 'Dynamic Value Average')
    ])

    principle = IntegerField('Principle Amount', [
        DataRequired('Principle Amount is required')
    ], widget=NumberInput(step=5, min=5))

    period = SelectField('Period', [
        DataRequired('Period is required'),
    ], choices=[
        ('daily', 'Daily'),
        ('weekly', 'Weekly'),
        ('monthly', 'Monthly'),
    ])

    apy = SelectField('Annual Percentage Yield', choices=[
        ('0.05', '5%'),
        ('0.075', '7.5%'),
        ('0.10', '10%'),
        ('0.125', '12.5%'),
        ('0.15', '15%')
    ])

    submit = SubmitField('Add Asset')
